### PR TITLE
Feature/add rabbit specific code

### PR DESCRIPTION
**Pull request by JamesGardiner**
This pull request introduces the bulk of the RabbitMQ specific sdx code to this new package. This is not feature complete yet - it specifically needs more unit testing - but getting eyes on at this point would be a good idea.

Significant changes from the existing implementations are that the classes now act as *real* package classes, with no dependencies on settings.py, and that individual services can now pass a processor function to the MessageConsumer class.

### CODE REVIEW

**Comment by JamesGardiner** [comment]
[(8) Create standalone rabbit library - Python](https://trello.com/c/suELjZyb/164-8-create-standalone-rabbit-library-python)

File: sdc/rabbit/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from .async_consumer import AsyncConsumer  # noqa
```
Comment (Dave-Haynes) [inline sdc/rabbit/__init__.py]: Can we absolute imports please?

File: sdc/rabbit/async_consumer.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
+import logging
+import time
+
+import pika
+from structlog import wrap_logger
+
+LOGGER = wrap_logger(logging.getLogger('__name__'))
+
+
+class AsyncConsumer(object):
```
Comment (Dave-Haynes) [inline sdc/rabbit/async_consumer.py]: No need to inherit from object.

File: sdc/rabbit/async_consumer.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+                LOGGER.error("Connection error", exception=e)
+                count += 1
+                LOGGER.error("Connection sleep", no_of_seconds=count)
+                time.sleep(count)
+
+                continue
+
+    def close_connection(self):
+        """This method closes the connection to RabbitMQ."""
+        LOGGER.info('Closing connection')
```
Comment (Dave-Haynes) [inline sdc/rabbit/async_consumer.py]: Prefer not using caps in variable names, especially for active objects.

File: sdc/rabbit/consumer.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
+import logging
+
+from structlog import wrap_logger
+
+from .exceptions import BadMessageError, DecryptError, RetryableError
+from .publisher import QueuePublisher
```
Comment (Dave-Haynes) [inline sdc/rabbit/consumer.py]: Absolute imports pls.

File: sdc/rabbit/test/test_consumer.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import json
+import logging
+import unittest
+
+from sdc.rabbit import AsyncConsumer, MessageConsumer
+
+
+class DotDict(dict):
```
Comment (Dave-Haynes) [inline sdc/rabbit/test/test_consumer.py]: This is so useful it could be in a module of its own for use by other tests.

File: sdc/rabbit/async_consumer.py (hunk 2 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: Agreed. This is taken straight from the AsyncConsumer in sdx-common.

File: sdc/rabbit/async_consumer.py (hunk 3 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: Done

File: sdc/rabbit/test/test_consumer.py (hunk 5 above)
Comment (JamesGardiner) [inline sdc/rabbit/test/test_consumer.py]: Agreed.

File: sdc/rabbit/__init__.py (hunk 1 above)
Comment (JamesGardiner) [inline sdc/rabbit/__init__.py]: Done - though not sure I agree that absolute imports are always better than explicit relative imports (though they are explicitly better than implicit relative imports :wink:).

File: sdc/rabbit/async_consumer.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+        :returns Object of type AsyncConsumer
+        :rtype AsyncConsumer
+
+        """
+        self._exchange = exchange
+        self._exchange_type = exchange_type
+        self._durable_queue = durable_queue
+        self._queue = rabbit_queue
+        self._rabbit_urls = rabbit_urls
+        self._rabbit_quarantine_queue = 'async_consumer_quarantine'
```
Comment (warrenbailey) [inline sdc/rabbit/async_consumer.py]: Why is this needed here?

File: sdc/rabbit/consumer.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            return None
+
+        try:
+            self._process(body.decode("utf-8"))
+            self._consumer.acknowledge_message(basic_deliver.delivery_tag,
+                                               tx_id=tx_id)
+
+        except DecryptError as e:
+            # Throw it into the quarantine queue to be dealt with
+            self._consumer.reject_message(basic_deliver.delivery_tag, tx_id=tx_id)
```
Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: Looks like its missing the Quarantine bit here too!!!

File: sdc/rabbit/async_consumer.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        self._channel.basic_nack(delivery_tag)
+
+    def reject_message(self, delivery_tag, **kwargs):
+        """Reject the message delivery from RabbitMQ by sending a
+        Basic.Reject RPC method for the delivery tag.
+        :param int delivery_tag: The delivery tag from the Basic.Deliver frame
+
+        """
+        logger.info('Rejecting message', delivery_tag=delivery_tag, **kwargs)
+        self._channel.basic_reject(delivery_tag, requeue=False)
```
Comment (warrenbailey) [inline sdc/rabbit/async_consumer.py]: Should we make the requeue True or at least optionall?

File: test_requirements.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+git+https://github.com/ONSDigital/sdx-common.git@0.7.0
+coverage==4.4.1
+responses==0.5.1
+pep8==1.6.2
+flake8==2.6.0
+structlog==17.2.0
```
Comment (warrenbailey) [inline test_requirements.txt]: structlog should be a runtime dependency, it's use in the code base, not just in tests. Please move to requirements.txt

File: sdc/rabbit/publisher.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class QueuePublisher(object):
+    """This is a queue publisher that publishes response messages to a
+    RabbitMQ queue.
+
+    """
+    _durable_queue = True
+
+    def __init__(self, urls, queue, **kwargs):
```
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: Remove **kwargs unless they are used

File: sdc/rabbit/publisher.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                                            delivery_mode=2
+                                        ),
+                                        body=message)
+            logger.debug("Published message")
+
+        except Exception as e:
+            logger.error("Unable to publish message", exception=repr(e))
+            return False
+        finally:
+            self._disconnect()
```
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: I still don't get the point of the disconnect here

File: sdc/rabbit/publisher.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                                        routing_key=self._queue,
+                                        properties=pika.BasicProperties(
+                                            content_type=content_type,
+                                            headers=headers,
+                                            delivery_mode=2
+                                        ),
+                                        body=message)
+            logger.debug("Published message")
+
+        except Exception as e:
```
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: I think we should have some more fine grain Exception handling here

File: sdc/rabbit/async_consumer.py [hunk 13]
```diff
@@ -0,0 +1,388 @@
+import logging
+import time
+
+import pika
+from structlog import wrap_logger
+
+logger = wrap_logger(logging.getLogger('__name__'))
+
+
+class AsyncConsumer():
```
Comment (warrenbailey) [inline sdc/rabbit/async_consumer.py]: Do you need the () ?

File: sdc/rabbit/consumer.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        is missing from message headers.
+
+
+        :param properties: Message properties
+
+        :returns: Delivery count value
+        :rtype: int
+
+        """
+        delivery_count = 0
```
Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: This ^ line isn't needed

File: sdc/rabbit/consumer.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+        """
+        self._consumer = consumer
+        self._process = process
+
+        if not callable(process):
+            msg = 'callback is not callable'
+            raise AttributeError(msg.format(process))
+        self._quarantine_publisher = QueuePublisher(self._consumer._rabbit_urls,
+                                                    self._consumer._rabbit_quarantine_queue)
```
Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: I don't like the access to the internal vars on the consumer and publisher. Feels wrong.

Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: What if someone changes it in the future without realising their accessed from outside the class

File: sdc/rabbit/async_consumer.py (hunk 8 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: I spoke to @necrophonic about this - the preference is to not requeue, but to quarantine in specific circumstances, otherwise we will enter a nacking cycle.

Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: I will add the quarantine code in though.

File: sdc/rabbit/consumer.py (hunk 7 above)
Comment (JamesGardiner) [inline sdc/rabbit/consumer.py]: Copy and pasted from sdx-common. I will re-add it.

File: sdc/rabbit/consumer.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+        :returns: None
+
+        """
+        try:
+            delivery_count = self.delivery_count(properties)
+        except KeyError as e:
+            self._consumer.reject_message(basic_deliver.delivery_tag)
+            msg = 'Bad message properties - no delivery count'
+            logger.error(msg, action="quarantined", exception=str(e))
```
Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: Log message is either lying or you need to quarantine the message here

File: sdc/rabbit/consumer.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                        queue=self._consumer._queue,
+                        delivery_tag=basic_deliver.delivery_tag,
+                        delivery_count=delivery_count,
+                        app_id=properties.app_id,
+                        tx_id=tx_id)
+
+        except KeyError as e:
+            self._consumer.reject_message(basic_deliver.delivery_tag)
+            logger.error("Bad message properties - no tx_id",
+                         action="quarantined",
```
Comment (warrenbailey) [inline sdc/rabbit/consumer.py]: and again!

File: sdc/rabbit/exceptions.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+class BadMessageError(Exception):
+    """A bad message is broken in some way that will never be accepted by
+    the endpoing and as such should be rejected (it will still be logged
+    and stored so no data is lost).
```
Comment (warrenbailey) [inline sdc/rabbit/exceptions.py]: Is that comment correct? Does it get logged and stored?

File: sdc/rabbit/async_consumer.py (hunk 8 above)
Comment (warrenbailey) [inline sdc/rabbit/async_consumer.py]: Ok, though if we forget to quarantine or someone accidentally removes that line it all goes to hell.

File: sdc/rabbit/async_consumer.py (hunk 6 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: It's not. I've taken this out and put a required parameter instead. This can be passed from the calling code so is configurable.

File: sdc/rabbit/publisher.py (hunk 10 above)
Comment (JamesGardiner) [inline sdc/rabbit/publisher.py]: :param **kwargs: Custom key/value pairs passed to the arguments
            parameter of pika's channel.queue_declare method

File: sdc/rabbit/publisher.py (hunk 11 above)
Comment (JamesGardiner) [inline sdc/rabbit/publisher.py]: Neither do I, it's probably ended up there because examples are scripts - connect, publish a message, disconnect... my preference is to leave this as technical debt for now - it's not efficient, but given the scale of other changes here I'd prefer to isolate issues.

File: sdc/rabbit/async_consumer.py (hunk 13 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: No

File: sdc/rabbit/consumer.py (hunk 14 above)
Comment (JamesGardiner) [inline sdc/rabbit/consumer.py]: Yep

File: sdc/rabbit/consumer.py (hunk 15 above)
Comment (JamesGardiner) [inline sdc/rabbit/consumer.py]: Yeah not great - I've changed so that the class expects a quarantine_publisher object to be explicitly passed.

File: sdc/rabbit/exceptions.py (hunk 18 above)
Comment (JamesGardiner) [inline sdc/rabbit/exceptions.py]: It depends on the `_process` callback, which is defined on a per-service basis. I will amend.

Comment (JamesGardiner) [inline sdc/rabbit/exceptions.py]: It is explicit in the docstrings that whatever is passed to Consumer as the process parameter, should raise these errors correctly:

Function or method to use for processsing message. Will be passed the body of the message as a string decoded using UTF - 8. Should raise sdc.rabbit.DecryptError, sdc.rabbit.BadMessageError or    
sdc.rabbit.RetryableError on failure, depending on the failure mode.

File: sdc/rabbit/async_consumer.py (hunk 8 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: message hell

File: sdc/rabbit/publisher.py (hunk 10 above)
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: Sorry missed this line 
self._arguments = kwargs

File: sdc/rabbit/publisher.py (hunk 11 above)
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: Yeah ok

File: sdc/rabbit/publisher.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+                                        properties=pika.BasicProperties(
+                                            content_type=content_type,
+                                            headers=headers,
+                                            delivery_mode=2
+                                        ),
+                                        body=message)
+            logger.debug("Published message")
+
+        except pika.exceptions.ChannelClosed:
+            self._connect()
```
Comment (warrenbailey) [inline sdc/rabbit/publisher.py]: should return false here

File: sdc/rabbit/publisher.py (hunk 10 above)
Comment (JamesGardiner) [inline sdc/rabbit/publisher.py]: :+1:

File: sdc/rabbit/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
+class BadMessageError(Exception):
+    """Raise if a message is broken in some way that means it will never
+    succesfully process.
+
+    """
+    pass
+
+
+class DecryptError(Exception):
```
Comment (necrophonic) [inline sdc/rabbit/exceptions.py]: `DecryptError` is certainly an error we use, but it's not strictly related to `rabbitmq` - it's use is very specific to the SDX decryption implementation and doesn't have wider usage.

Maybe in the context of the standalone lib it should be a `QuarantinableError` or similar (as that's what it really refers to in operation, whether that was caused by a bad decrypt or other "hold it for now; it looks a bit dodgy" error)?

File: Makefile [hunk 21]
```diff
@@ -0,0 +1,17 @@
+
+.PHONY: depinstall clean sdist test
```
Comment (necrophonic) [inline Makefile]: The `README` doesn't correctly reference actual targets in the `Makefile`

Should also have a sensible `make` default. Normal behaviour should be along the lines of:
```
$ make
$ make test
$ make install
```

The `make (all)` should pull in any dependencies etc that it needs, the `test` should, well, run the tests, and the more optional `install` should potentially `pip` install the lib to the local env.

https://www.gnu.org/prep/standards/html_node/Standard-Targets.html

File: Makefile [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+clean:
+	rm -rf sdc/rabbit/doc/html
+	rm -v dist/sdc-rabbit-python-*.tar.gz
+
+depinstall:
+	pip3 install -r requirements.txt
+
+sdist:
```
Comment (necrophonic) [inline Makefile]: Strictly an `sdist` in the python world, but the step may be more commonly understood as `dist`

File: sdc/rabbit/async_consumer.py [hunk 23]
```diff
@@ -0,0 +1,387 @@
+import logging
```
Comment (necrophonic) [inline sdc/rabbit/async_consumer.py]: Is there a way to pass in logging formatting et al to the library (without relying on `sdx-common` which a consumer may not want to also pull in just for logging format and they may have a differing requirement)?

File: README.md [hunk 24]
```diff
@@ -2,7 +2,7 @@
...
+[![Build Status](https://travis-ci.org/ONSdigital/sdc-rabbit-python.svg?branch=master)](https://travis-ci.org/ONSdigital/sdc-rabbit-python) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/043810e79dac47759cc661361a8af12b)](https://www.codacy.com/app/ons-sdc/sdc-rabbit-python?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ONSdigital/sdc-rabbit-python&amp;utm_campaign=Badge_Grade) [![codecov](https://codecov.io/gh/ONSdigital/sdc-rabbit-python/branch/develop/graph/badge.svg)](https://codecov.io/gh/ONSdigital/sdc-rabbit-python)
 
 Common code for SDC Pika-based services that interact with RabbitMQ.
```
Comment (necrophonic) [inline README.md]: `README` should also give more of a run down of what the lib offers and the options it has available (also listing any assumptions made/restrictions - like "all queues are assumed to be durable")

File: sdc/rabbit/async_consumer.py (hunk 23 above)
Comment (JamesGardiner) [inline sdc/rabbit/async_consumer.py]: Done - sdx-common was only getting pulled in to format the pika logging in the unit tests. It is no longer required.
